### PR TITLE
feat(parameters): add parameter set commands

### PR DESCRIPTION
**Pull request by vdice**
# What does this change
 * Adds Parameter Set generation cmds/logic
 * Refactors previous generation logic into common code and credential-/parameter-specific logic

# What issue does it fix
Closes https://github.com/deislabs/porter/issues/1073
Contributes towards https://github.com/deislabs/porter/issues/1053

# Notes for the reviewer
N/A

# Checklist
- [x] Unit Tests
- [x] Documentation
- [x] Schema (porter.yaml) N/A

### CODE REVIEW

File: pkg/generator/generator.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	sourceValuePromptTemplate := "Enter the %s that will be used to set %s %q\n "
+
+	c := valuesource.Strategy{Name: name}
+
+	source := ""
+	if err := survey.AskOne(sourceTypePrompt, &source, nil); err != nil {
+		return c, err
+	}
+
+	// TODO: add 'skip' option *if* surveyType = surveyParameters *and* parameter not required
```
Comment (vdice) [inline pkg/generator/generator.go:66]: Thoughts on these 2 TODOs?  Tackle in this PR?  Follow-ups?  (Assuming functionality is desired...)

File: pkg/porter/parameters.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+		var rows [][]string
+
+		// Iterate through all ParameterStrategies and add to rows
+		for _, pset := range paramSet.Parameters {
+			sourceVal, sourceType := GetParameterSourceValueAndType(pset.Source)
+			rows = append(rows, []string{pset.Name, sourceVal, sourceType})
+		}
+
+		// Build and configure our tablewriter
+		table := tablewriter.NewWriter(p.Out)
```
Comment (carolynvs) [inline pkg/porter/parameters.go:238]: Can we make a reusable function to initialize this table here any in any other places that use this?

File: pkg/porter/parameters.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+		return nil
+	default:
+		return fmt.Errorf("invalid format: %s", opts.Format)
+	}
+}
+
+// GetParameterSourceValueAndType takes a given valuesource.Source struct and
+// returns the source value itself as well as source type, e.g., 'path', 'env', etc.,
+// both in their string forms
+func GetParameterSourceValueAndType(pset valuesource.Source) (value string, key string) {
```
Comment (carolynvs) [inline pkg/porter/parameters.go:266]: Seems like this doesn't need to be a function anymore (same as `GetCredentialSourceValueAndType`). Now that we have a common `valuesource.Source` and this just operates on its fields, it doesn't do anything that couldn't just use the fields directly.

File: pkg/porter/parameters.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+// ParameterDeleteOptions represent options for Porter's parameter delete command
+type ParameterDeleteOptions struct {
+	Name string
+}
+
+// DeleteParameter deletes the parameter set corresponding to the provided
+// names.
+func (p *Porter) DeleteParameter(opts ParameterDeleteOptions) error {
+	err := p.Parameters.Delete(opts.Name)
+	if err == crud.ErrRecordDoesNotExist {
```
Comment (carolynvs) [inline pkg/porter/parameters.go:279]: I've found that with plugins the check should be this instead

```suggestion
	if err != nil && strings.Contains(err.Error(), crud.ErrRecordDoesNotExist) {
```

File: pkg/porter/parameters.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+		if p.Debug {
+			fmt.Fprintln(p.Err, "parameter set does not exist")
+		}
+		return nil
+	}
+	return errors.Wrapf(err, "unable to delete parameter")
+}
+
+// Validate validates the args provided Porter's parameter delete command
```
Comment (carolynvs) [inline pkg/porter/parameters.go:288]:
```suggestion
// Validate the args provided to the delete parameter command
```

File: pkg/porter/parameters.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+		return err
+	}
+	o.Name = args[0]
+	return nil
+}
+
+func validateParameterName(args []string) error {
+	switch len(args) {
+	case 0:
+		return errors.Errorf("no parameter name was specified")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:300]:
```suggestion
		return errors.Errorf("no parameter set name was specified")
```

File: pkg/porter/parameters.go [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+	case 1:
+		return nil
+	default:
+		return errors.Errorf("only one positional argument may be specified, the parameter name, but multiple were received: %s", args)
```
Comment (carolynvs) [inline pkg/porter/parameters.go:304]:
```suggestion
		return errors.Errorf("only one positional argument may be specified, the parameter set name, but multiple were received: %s", args)
```

File: pkg/porter/parameters.go [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	return g.bundleFileOptions.Validate(cxt)
+}
+
+func (g *ParameterOptions) validateParamName(args []string) error {
+	if len(args) == 1 {
+		g.Name = args[0]
+	} else if len(args) > 1 {
+		return errors.Errorf("only one positional argument may be specified, the parameter name, but multiple were received: %s", args)
```
Comment (carolynvs) [inline pkg/porter/parameters.go:90]:
```suggestion
		return errors.Errorf("only one positional argument may be specified, the parameter set name, but multiple were received: %s", args)
```

File: pkg/porter/parameters.go [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+		name = bundle.Name
+	}
+	genOpts := generator.GenerateParametersOptions{
+		GenerateOptions: generator.GenerateOptions{
+			Name:   name,
+			Silent: opts.Silent,
+		},
+		Parameters: bundle.Parameters,
+	}
+	fmt.Fprintf(p.Out, "Generating new parameter %s from bundle %s\n", genOpts.Name, bundle.Name)
```
Comment (carolynvs) [inline pkg/porter/parameters.go:128]:
```suggestion
	fmt.Fprintf(p.Out, "Generating new parameter set %s from bundle %s\n", genOpts.Name, bundle.Name)
```

File: pkg/porter/parameters.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+	fmt.Fprintf(p.Out, "==> %d parameters declared for bundle %s\n", len(genOpts.Parameters), bundle.Name)
+
+	pset, err := generator.GenerateParameters(genOpts)
+	if err != nil {
+		return errors.Wrap(err, "unable to generate parameters")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:133]:
```suggestion
		return errors.Wrap(err, "unable to generate parameter set")
```

File: pkg/porter/parameters.go [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+	if opts.DryRun {
+		data, err := json.Marshal(pset)
+		if err != nil {
+			return errors.Wrap(err, "unable to generate parameters JSON")
+		}
+		fmt.Fprintf(p.Out, "%v", string(data))
+		return nil
+	}
+	err = p.Parameters.Save(*pset)
+	return errors.Wrapf(err, "unable to save parameters")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:148]:
```suggestion
	return errors.Wrapf(err, "unable to save parameter set")
```

File: pkg/porter/parameters.go [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+// EditParameter edits the parameters of the provided name.
+func (p *Porter) EditParameter(opts ParameterEditOptions) error {
+	paramSet, err := p.Parameters.Read(opts.Name)
+	if err != nil {
+		return err
+	}
+
+	contents, err := yaml.Marshal(paramSet)
+	if err != nil {
+		return errors.Wrap(err, "unable to load parameters")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:178]:
```suggestion
		return errors.Wrap(err, "unable to load parameter set")
```

File: pkg/porter/parameters.go [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	editor := editor.New(p.Context, fmt.Sprintf("porter-%s.yaml", paramSet.Name), contents)
+	output, err := editor.Run()
+	if err != nil {
+		return errors.Wrap(err, "unable to open editor to edit parameters")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:184]:
```suggestion
		return errors.Wrap(err, "unable to open editor to edit parameter set")
```

File: pkg/porter/parameters.go [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	err = yaml.Unmarshal(output, &paramSet)
+	if err != nil {
+		return errors.Wrap(err, "unable to process parameters")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:189]:
```suggestion
		return errors.Wrap(err, "unable to process parameter set")
```

File: pkg/porter/parameters.go [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	err = p.Parameters.Validate(paramSet)
+	if err != nil {
+		return errors.Wrap(err, "parameters are invalid")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:194]:
```suggestion
		return errors.Wrap(err, "parameter set is invalid")
```

File: pkg/porter/parameters.go [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+	}
+
+	paramSet.Modified = time.Now()
+	err = p.Parameters.Save(paramSet)
+	if err != nil {
+		return errors.Wrap(err, "unable to save parameters")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:200]:
```suggestion
		return errors.Wrap(err, "unable to save parameter set")
```

File: pkg/porter/parameters.go [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+// names.
+func (p *Porter) DeleteParameter(opts ParameterDeleteOptions) error {
+	err := p.Parameters.Delete(opts.Name)
+	if err == crud.ErrRecordDoesNotExist {
+		if p.Debug {
+			fmt.Fprintln(p.Err, "parameter set does not exist")
+		}
+		return nil
+	}
+	return errors.Wrapf(err, "unable to delete parameter")
```
Comment (carolynvs) [inline pkg/porter/parameters.go:285]:
```suggestion
	return errors.Wrapf(err, "unable to delete parameter set")
```

File: cmd/porter/parameters.go [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVarP(&opts.File, "file", "f", "",
+		"Path to the porter manifest file. Defaults to the bundle in the current directory.")
+	f.StringVar(&opts.CNABFile, "cnab-file", "",
+		"Path to the CNAB bundle.json file.")
+	f.BoolVar(&opts.DryRun, "dry-run", false,
+		"Generate parameter but do not save it.")
```
[the remark this answers is not in the document]
Comment (vdice) [reply cmd/porter/parameters.go:83]: Good call, I'd be in support of this.

File: pkg/porter/parameters.go (hunk 2 above)
Comment (vdice) [reply pkg/porter/parameters.go:237]: Good call; created a follow-up PR from this comment.

**Comment by vdice** [comment]
Thank you for all of the great feedback @carolynvs , especially the detailed docs updates.  I believe I've addressed them all.  Please take another look, when convenient.

File: pkg/porter/parameters.go [hunk 19]
```diff
@@ -276,7 +268,7 @@ type ParameterDeleteOptions struct {
 // names.
 func (p *Porter) DeleteParameter(opts ParameterDeleteOptions) error {
 	err := p.Parameters.Delete(opts.Name)
-	if err != nil && strings.Contains(err.Error(), crud.ErrRecordDoesNotExist) {
+	if err != nil && strings.Contains(err.Error(), crud.ErrRecordDoesNotExist.Error()) {
```
Comment (carolynvs) [inline pkg/porter/parameters.go:271]: Oops! Sorry about that, gave you the wrong code... 😬

**Pull request merged**